### PR TITLE
Review

### PR DESCRIPTION
**Pull request by Neverlord**

### CODE REVIEW

File: caf/crdt/atom_types.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+namespace caf {
+namespace crdt {
+
+// -------- Public usable atoms ------------------------------------------------
+
+/// Send to subscribed actors when a replica has changed
+using notify_atom = atom_constant<atom("notify")>;
+
+/// Send to replicator to read states from a majority of nodes
+using read_majority_atom = atom_constant<atom("readmajor")>;
```
Comment (Neverlord) [inline caf/crdt/atom_types.hpp:35]: All lowercase is hard to read. We generally use `camelCase` for atom strings composed of multiple words. For example [`getName`](https://github.com/actor-framework/actor-framework/blob/272669fabce47763b38f2a55cf05ef479f4eb4a7/libcaf_core/test/composable_behavior.cpp#L162) rather than `getname`.

File: caf/crdt/atom_types.hpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+/// @private
+using tick_state_atom = atom_constant<atom("fstate")>;
+
+/// @private
+using tick_topics_atom = atom_constant<atom("ftopics")>;
+
+/// @private
+using shutdown_atom = atom_constant<atom("shutdown")>;
+
+// -------- Replicator communication atoms -------------------------------------
```
Comment (Neverlord) [inline caf/crdt/atom_types.hpp:78]: Replicator communication atoms are internal as well, no? The distinction makes little sense to me.

File: caf/crdt/all.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
+/******************************************************************************
+ *                       ____    _    _____                                   *
+ *                      / ___|  / \  |  ___|    C++                           *
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright (C) 2011 - 2016                                                  *
+ * Dominik Charousset <dominik.charousset (at) haw-hamburg.de>                *
+ * Marian Triebe <marian.triebe (at) haw-hamburg.de>                          *
```
Comment (Neverlord) [inline caf/crdt/all.hpp:10]: The copyright note at the top of each header refers to CAF itself, which your module is a part of. Your authorship is stored in each file you have contributed to. This is why GitHub lists contributors on a per-file level. No reason to alter the CAF copyright note.

File: caf/crdt/crdt_config.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "caf/actor_system_config.hpp"
+
+#include "caf/crdt/detail/replica.hpp"
+
+namespace caf {
+namespace crdt {
+
+/// Extended `actor_system_config` to support the CRDT module.
+struct crdt_config : public virtual actor_system_config {
```
Comment (Neverlord) [inline caf/crdt/crdt_config.hpp:32]: - Misuse of `struct` as class. This is symptomatic for the entire code base. Please use `struct` *only* for modeling POD (plain-old-data) containers. Once you use inheritance, member functions, or private fields please use a `class`. I only add a comment here, but there are many mis-used `struct`s. Being lazy and saving one `public:` at the top of your class is not a reason to use `struct`.
- Why did you use virtual inheritance here? A config class is not supposed to act as compositional building block.

File: caf/crdt/detail/distribution_layer.hpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  /// Add a freshly discovered node
+  /// @param nid node to add
+  void add_new_node(const node_id& nid) {
+    auto& mm = impl_->home_system().middleman();
+    auto query = mm.remote_lookup(replicator_atom::value, nid);
+    if (query) {
+      auto repl = actor_cast<replicator_actor>(query);
+      store_.emplace(nid, std::make_tuple(0, repl, std::unordered_set<uri>{}));
+      send_as(impl_, repl, get_topics_atom::value, size_t{0});
+    }
```
Comment (Neverlord) [inline caf/crdt/detail/distribution_layer.hpp:60]: No error handling? Is there a reason not to return `error` from this function and leave it up to the user to resolve connection issues?

File: caf/crdt/detail/distribution_layer.hpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    modify_topics(u, false);
+  }
+
+  /// Locally remove a topic
+  /// @param u uri to remove
+  void remove_topic(const uri& u) {
+    modify_topics(u, true);
+  }
+
+  /// Send the last seen version of a node to the node, if there
```
Comment (Neverlord) [inline caf/crdt/detail/distribution_layer.hpp:96]: Who is "the node"? All remote nodes?

File: caf/crdt/detail/distribution_layer.hpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  /// is a newer version, the node will respond with its updated data.
+  void pull_topics() {
+    for (auto& entry : store_) {
+      if (entry.first == impl_->node())
```
Comment (Neverlord) [inline caf/crdt/detail/distribution_layer.hpp:100]: This `if (...) continue;` guard is used in almost all places where `store_` is used. Why not store local bookkeeping data separately instead and have the map contain only entries for remote nodes?

File: caf/crdt/detail/distribution_layer.hpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+#include <unordered_set>
+
+namespace caf {
+namespace crdt {
+namespace detail {
+
+/// Manages the data send between nodes.
+struct distribution_layer {
+
+  using tuple_type = std::tuple<size_t, replicator_actor,
```
Comment (Neverlord) [inline caf/crdt/detail/distribution_layer.hpp:39]: Using a tuple here makes it hard to follow the code below. Why not have a struct and give the fields meaningful names instead of `get<0>(...)`?

A struct like this would make the code more readable:

```
struct node_data {
  size_t version;
  replicator_actor replicator;
  std::unordered_set<uri> filter;
};
```

File: caf/crdt/detail/register_types.hpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+namespace crdt {
+namespace detail {
+
+struct register_types {
+
+  register_types(actor_system_config& cfg) : cfg_(cfg) {
+    // nop
+  }
+
+  void operator()() noexcept {
```
Comment (Neverlord) [inline caf/crdt/detail/register_types.hpp:42]: What is the purpose of this struct? I see absolutely no benefit over a free function here.

File: caf/crdt/detail/replica.hpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+#include "caf/crdt/notifyable.hpp"
+#include "caf/crdt/replicator_actor.hpp"
+
+#include <unordered_set>
+
+namespace caf {
+namespace crdt {
+namespace detail {
+
+namespace {
```
Comment (Neverlord) [inline caf/crdt/detail/replica.hpp:37]: Anonymous namespace inside a header means all symbols are duplicated for each compilation unit that includes this header. Is this really the intention here?

File: caf/crdt/detail/replica.hpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  size_t messages_left;
+  T crdt;
+  actor respond_to;
+  response_promise rp;
+};
+
+///
+template <class T>
+behavior writer(stateful_actor<wr_state<T>>* self) {
+  auto init = [=](size_t msgs_left, actor respond) {
```
Comment (Neverlord) [inline caf/crdt/detail/replica.hpp:51]: Why pass the actor as argument to the function if all instances calling this function pass `self->current_sender()` anyway?

File: caf/crdt/detail/replica.hpp [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+      if (--self->state.messages_left == 0)
+        self->state.rp.deliver(write_succeed_atom::value);
+    }
+  };
+}
+
+///
+template <class T>
+behavior reader(stateful_actor<wr_state<T>>* self) {
+  auto init = [=](size_t msgs_left, actor respond) {
```
Comment (Neverlord) [inline caf/crdt/detail/replica.hpp:84]: Same here: `respond == self->current_sender()` on all paths.

File: caf/crdt/detail/replica.hpp [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+///
+template <class T>
+struct replica : public event_based_actor {
+  replica(actor_config& cfg, const uri& topic)
+      : event_based_actor(cfg), topic_{topic} {
+    // nop
+  }
+
+protected:
+  virtual behavior make_behavior() override {
```
Comment (Neverlord) [inline caf/crdt/detail/replica.hpp:124]: Don't add `virtual` if the function is already marked as `override`.

File: caf/crdt/detail/replicator_hooks.hpp [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+#include "caf/scoped_actor.hpp"
+
+#include "caf/io/hook.hpp"
+
+namespace caf {
+namespace crdt {
+namespace detail {
+
+/// IO-Hooks used by replicator
+class replicator_hooks : public io::hook {
```
Comment (Neverlord) [inline caf/crdt/detail/replicator_hooks.hpp:33]: The base class uses singular. Why switch to plural here?

File: caf/crdt/lamport_clock.hpp [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  /// Merge two lamport clocks
+  lamport_clock merge(const lamport_clock& other) {
+    time_ = std::max(time_, other.time_);
+    return {time_};
+  }
+
+  /// @returns the current time
+  uint64_t time() const { return time_; }
```
Comment (Neverlord) [inline caf/crdt/lamport_clock.hpp:29]: This accesses the value. Why not use `operator*`?

File: caf/crdt/lamport_clock.hpp [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+  /// @returns post-increment internal time and returns its value
+  uint64_t increment() { return ++time_; }
```
Comment (Neverlord) [inline caf/crdt/lamport_clock.hpp:32]: Why not use `operator++`?

File: caf/crdt/lamport_clock.hpp [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+  /// @cond PRIVATE
+
+  intptr_t compare(uint64_t lhs, uint64_t rhs) const noexcept {
```
Comment (Neverlord) [inline caf/crdt/lamport_clock.hpp:36]: Can be `static` + `private`.

File: caf/crdt/notifyable.hpp [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+#include "caf/crdt/atom_types.hpp"
+
+namespace caf {
+namespace crdt {
+
+/// Interface definition for actors which work with CRDT States and support
+/// notifications.
+template <class State>
+using notifyable = typed_actor<
```
Comment (Neverlord) [inline caf/crdt/notifyable.hpp:34]: From the [style guide](https://github.com/actor-framework/actor-framework/blob/master/CONTRIBUTING.md#template-metaprogramming):

> Brake `using name = ...` statements always directly after `=` if it does not fit in one line.

Since this does fit into one line:

```cpp
template <class State>
using notifyable = typed_actor<reacts_to<notify_atom, State>>;
```

Otherwise, this is the preferred CAF style:

```cpp
template <class State>
using foo =
  typed_actor<
    reacts_to<bar, State>,
    reacts_to<baz, State>
  >;
```

File: caf/crdt/replicator.hpp [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+protected:
+  replicator(actor_system&);
+  ~replicator();
+
+private:
+  actor_system& system_;
+  replicator_actor manager_;
+
+public:
```
Comment (Neverlord) [inline caf/crdt/replicator.hpp:68]: Please stay consistent. The order is: `public`, `protected`, and finally `private`.

File: caf/crdt/replicator_actor.hpp [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+#include "caf/crdt/uri.hpp"
+#include "caf/crdt/atom_types.hpp"
+
+#include <unordered_set>
+
+namespace caf {
+namespace crdt {
+
+/// Interface of replicator
+using replicator_actor = typed_actor<
```
Comment (Neverlord) [inline caf/crdt/replicator_actor.hpp:38]: Break after `=`: https://github.com/actor-framework/actor-framework/blob/master/CONTRIBUTING.md#template-metaprogramming

File: caf/crdt/types/base_datatype.hpp [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    // If owner_ is no longer valid, the actor system got shut down and it is
+    // no longer possible nor necessary to unsubscribe
+    if (owner_) {
+      auto hdl = owner_.home_system().replicator().actor_handle();
+      send_as(owner_, hdl, unsubscribe_atom::value, uri{topic_});
+    }
+  }
+
+  /// @returns topic of this state
+  inline std::string topic() const { return topic_.str(); }
```
Comment (Neverlord) [inline caf/crdt/types/base_datatype.hpp:59]: Is topic really the right term here? From the code this looks like an identifier to me.

File: caf/crdt/types/base_datatype.hpp [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+
+  /// @returns the owner of this state
+  inline const actor& owner() const { return owner_; }
+
+  /// @private
+  template <class Processor>
+  friend void serialize(Processor& proc, base_datatype& x) {
```
Comment (Neverlord) [inline caf/crdt/types/base_datatype.hpp:66]: Why is this serializable if only half the fields are written?

File: caf/crdt/types/gcounter.hpp [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  ///
+  void increment() {
+    increment_by(1);
+  }
+
+  ///
+  inline T count() const {
+    T result = 0;
```
Comment (Neverlord) [inline caf/crdt/types/gcounter.hpp:75]: Please don't implement well-known algorithms by hand: http://en.cppreference.com/w/cpp/algorithm/accumulate

File: caf/crdt/types/gcounter.hpp [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+    for (auto& elem : map_)
+      result += elem.second;
+    return result;
+  }
+
+  /// Merge function, for this type it is simple
```
Comment (Neverlord) [inline caf/crdt/types/gcounter.hpp:81]: Comment, for this function is useless

File: caf/crdt/types/gcounter.hpp [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+  /// @param other delta-CRDT to merge into this
+  /// @returns a delta gcounter<T>
+  gcounter<T> merge(const gcounter<T>& other) {
+    std::unordered_map<actor, T> delta;
+    for (auto& elem : other.map_) {
+      auto key = elem.first;
```
Comment (Neverlord) [inline caf/crdt/types/gcounter.hpp:87]: `auto& key`

File: caf/crdt/types/gcounter.hpp [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+      if (map_[key] < elem.second) {
+        map_[key] = elem.second;
```
Comment (Neverlord) [inline caf/crdt/types/gcounter.hpp:89]: Second occurrence of `map_[key]` -> two `O(log n)` lookups for the exact same element.

File: caf/crdt/types/gmap.hpp [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+
+public:
+
+  gmap merge(const gmap& other) {
+    Container delta;
+    for (auto& entry : other.map_) {
+      auto& key = entry.first;
+      auto& value = entry.second;
+      if (map_[key] < value) {
+        map_[key] = value;
```
Comment (Neverlord) [inline caf/crdt/types/gmap.hpp:68]: Repeated lookup.

File: caf/crdt/types/gmap.hpp [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+      return true;
+    }
+    return false;
+  }
+
+  ///
+  optional<Value> get(const Key& key) const {
+    auto iter = map_.find(key);
+    if (iter == map_.end())
+      return empty;
```
Comment (Neverlord) [inline caf/crdt/types/gmap.hpp:93]: `return none;`

File: caf/crdt/types/gmap.hpp [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+
+private:
+  void internal_assign(const Key& key, const Value& value) {
+    map_[key] = value;
+    publish(gmap{key, value});
+  }
+
+  Container map_;  /// Internal map
+
+public:
```
Comment (Neverlord) [inline caf/crdt/types/gmap.hpp:119]: Public functions at bottom of class.

File: caf/crdt/types/lww_register.hpp [hunk 30]
```diff
@@ -0,0 +1,251 @@
...
+#ifndef CAF_CRDT_TYPES_LWW_REGISTER_HPP
+#define CAF_CRDT_TYPES_LWW_REGISTER_HPP
+
+#include "caf/node_id.hpp"
+
+#include "caf/crdt/lamport_clock.hpp"
+
+#include "caf/crdt/types/base_datatype.hpp"
+
+// TODO: Add unit test for this type!
```
Comment (Neverlord) [inline caf/crdt/types/lww_register.hpp:30]: Commenting out whole files litters the code base. Simply remove from the repository entirely. Git allows you to dig out old files easily once you continue working on this.

File: caf/crdt/uri.hpp [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+#define CAF_CRDT_URI_HPP
+
+#include "caf/actor_system.hpp"
+#include "caf/intrusive_ptr.hpp"
+
+#include "caf/detail/comparable.hpp"
+
+namespace {
+
+constexpr char wildcard = '*';
```
Comment (Neverlord) [inline caf/crdt/uri.hpp:31]: This pollutes the global namespace with a very common name.

File: caf/crdt/uri.hpp [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+constexpr char path_delim = '/';
+
+} // namespace <anonymous>
+
+namespace caf {
+namespace crdt {
+
+/// Uri implementation, that supports a scheme and a path.
+class uri : public caf::detail::comparable<uri> {
```
Comment (Neverlord) [inline caf/crdt/uri.hpp:40]: `comparable` is usually inherited privately.

File: caf/crdt/uri.hpp [hunk 33]
```diff
@@ -0,0 +1,170 @@
...
+
+  enum class states {
+    parse_scheme,
+    parse_slash,
+    parse_topic,
+    parse_end
+  } state_;
+
+  bool consume(char c) {
+    // -----
```
Comment (Neverlord) [inline caf/crdt/uri.hpp:95]: // ?????

File: caf/crdt/vector_clock.hpp [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+
+#include <unordered_map>
+
+namespace caf {
+namespace crdt {
+
+/// Return type for vector clock compares
+/// @relates `vector_clock`
+enum vector_clock_result {
+  GREATER,
```
Comment (Neverlord) [inline caf/crdt/vector_clock.hpp:34]: WE DO NOT USE UPPERCASE LETTERS FOR ENUMERATIONS. ONLY FOR MACROS AND ONLY TO MAKE MACROS LOOK SILLY AND UGLY BECAUSE THEY ARE. OR DID YOU JUST FEEL LIKE SCREAMING?

File: caf/crdt/vector_clock.hpp [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+  EQUAL,
+  SMALLER,
+  SIMULTANEOUS
```
Comment (Neverlord) [inline caf/crdt/vector_clock.hpp:37]: The usual term used in vector clocks is concurrent, no?

File: examples/distributed.cpp [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+class incrementer : public types::gcounter<int>::base {
+public:
+  incrementer(actor_config& cfg)
+    : types::gcounter<int>::base(cfg),
+      state_(this, "gcounter<int>://counter") {
+    // nop
+  }
+
+protected:
+  typename types::gcounter<int>::behavior_type make_behavior() override {
```
Comment (Neverlord) [inline examples/distributed.cpp:48]: `typename` is only needed in templates.

File: examples/distributed.cpp [hunk 37]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+class config : public crdt_config {
+public:
+  config() {
+    add_crdt<types::gcounter<int>>("gcounter<int>");
+  }
+};
+
+void caf_main(actor_system& system, const config&) {
```
Comment (Neverlord) [inline examples/distributed.cpp:72]: This looks like a unit test to me, not like a best practice example.

File: examples/parse_uri.cpp [hunk 38]
```diff
@@ -0,0 +1,69 @@
...
+  auto u = uri{"gset<int>://rand"};
+  std::cout << std::boolalpha << "assumed (true, false): "
+            << u.match_rtti<types::gset<int>>(system) << ", "
+            << u.match_rtti<types::gset<float>>(system) << std::endl;
+  return 0;
+}
+*/
+
+int main() {
+
```
Comment (Neverlord) [inline examples/parse_uri.cpp:68]: Please remove empty files.

File: examples/simple.cpp [hunk 39]
```diff
@@ -0,0 +1,49 @@
...
+class config : public crdt_config {
+public:
+  config() {
+    add_crdt<types::gset<int>>("gset<int>");
+  }
+};
+
+void actor_fun(event_based_actor* self) {
+  // Initialize state
+  auto state_ = std::make_shared<gset<int>>(self, "gset<int>://set");
```
Comment (Neverlord) [inline examples/simple.cpp:24]: Using `make_shared` adds an unnecessary heap indirection. Using a stateful actor here instead is more efficient and idiomatic.

File: examples/worker.cpp [hunk 40]
```diff
@@ -0,0 +1,195 @@
...
+
+#include "caf/crdt/lamport_clock.hpp"
+
+using namespace std;
+using namespace caf;
+using namespace caf::crdt;
+
+namespace std {
+
+template <>
```
Comment (Neverlord) [inline examples/worker.cpp:41]: There is no reason to implement this as a template. Plus, there's also no reason to repeat `std::` in `namespace std`. Finally, monkey-patching `operator<` in `namespace std` with a very strange semantics (only considering the first half of a pair) is downright evil. All containers in the C++ standard library allow you do specify a comparator.

File: src/replicator.cpp [hunk 41]
```diff
@@ -0,0 +1,87 @@
...
+using namespace caf::crdt;
+
+void replicator::start() {
+  manager_ = make_replicator_actor(system_);
+  system_.registry().put(replicator_atom::value,
+                         actor_cast<strong_actor_ptr>(manager_));
+}
+
+void replicator::stop() {
+  system_.registry().erase(replicator_atom::value);
```
Comment (Neverlord) [inline src/replicator.cpp:46]: The registry is cleared up on shutdown anyway.

File: src/replicator_actor.cpp [hunk 42]
```diff
@@ -0,0 +1,146 @@
...
+        delegate(find_actor(topic), publish_atom::value, msg);
+      },
+      [&](tick_state_atom) {
+        // All states have to send their state to the replicator
+        for (auto& state : states_)
+          anon_send(state.second, copy_atom::value);
+        delayed_send(this, std::chrono::seconds(1), tick_state_atom::value);
+      },
+      [&](copy_ack_atom, const uri& u, const message& msg) {
+        dist.publish(u, msg);
```
Comment (Neverlord) [inline src/replicator_actor.cpp:70]: Can't we move `u` and `msg` here?

File: src/vector_clock.cpp [hunk 43]
```diff
@@ -0,0 +1,100 @@
...
+  if (greater && !smaller) return GREATER;
+  if (!greater && smaller) return SMALLER;
+  return SIMULTANEOUS;
+}
+
+vector_clock vector_clock::merge(const vector_clock& other) {
+  map_type delta;
+  // Copy entries into delta which are not in `this`
+  auto not_in_other = [&](const value_type& e) {
+    return other.map_.find(e.first) == other.map_.end();
```
Comment (Neverlord) [inline src/vector_clock.cpp:75]: A bit more compact: `other.map_.count(e.first) == 0`.

**Review by Neverlord — COMMENTED** [review]
Documentation is lackluster, but overall design seems ok at first glance. One thing I didn't get is why topics aren't called IDs. From what I can tell a topic uniquely identifies a data set / counter.

File: caf/crdt/detail/distribution_layer.hpp (hunk 5 above)
Comment (Hamdor) [reply caf/crdt/detail/distribution_layer.hpp:60]: There is no error handling because a node only has an replicator if the crdt module is loaded. If we discover a node without replicator, we ignore it

File: caf/crdt/types/gmap.hpp (hunk 29 above)
Comment (Hamdor) [reply caf/crdt/types/gmap.hpp:119]: the problem with these declarations is, that `decltype` can only be used after declaration of `map_`...

File: examples/simple.cpp (hunk 39 above)
Comment (Hamdor) [reply examples/simple.cpp:24]: The problem is, that we can't pass arguments to the constructor of the state (if statefull_actor is used)

File: caf/crdt/detail/distribution_layer.hpp (hunk 5 above)
Comment (Neverlord) [reply caf/crdt/detail/distribution_layer.hpp:60]: If we discover a node without replicator that might be a configuration error of the user or something else going really wrong. It might also be that node disconnected and we could consider trying to re-connect periodically.

File: caf/crdt/types/gmap.hpp (hunk 29 above)
Comment (Neverlord) [reply caf/crdt/types/gmap.hpp:119]: There's no reason to use decltype in the first place. You already have const_iterator defined and used above.

File: examples/simple.cpp (hunk 39 above)
Comment (Neverlord) [reply examples/simple.cpp:24]: You don't pass any arguments to the ctor of your gset other than `self` and a constant. And the state is allowed to accept a `self` pointer of type `event_based_actor*` in its ctor.
